DXMainClient/Online/Connection.cs

```diff
@@ -220,70 +220,59 @@ private void ConnectToServer()
-            int bytesRead;
 
             Register();
 
-            Timer timer = new Timer(new TimerCallback(AutoPing), null, 30000, 120000);
+            Timer timer = new Timer(AutoPing, null, 30000, 120000);
 
             connectionCut = true;
 
             while (true)
             {
-                bytesRead = 0;
-
                 if (connectionManager.GetDisconnectStatus())
                 {
                     connectionManager.OnDisconnected();
                     connectionCut = false; // This disconnect is intentional
                     break;
                 }
 
+                if (!serverStream.DataAvailable)
+                {
+                    Thread.Sleep(10);
+                    continue;
+                }
+
+                int bytesRead;
+
                 try
                 {
                     bytesRead = serverStream.Read(message, 0, 1024);
                 }
-                catch (Exception ex)
+                catch (IOException ex) when (IsKnownSocketException(ex))
                 {
                     errorTimes++;
-
-                    if (errorTimes > 30) // TODO Figure out if this hacky check is actually necessary
-                    {
-                        Logger.Log("Disconnected from CnCNet due to a socket error. Message: " + ex.Message);
-                        failedServerIPs.Add(currentConnectedServerIP);
-                        connectionManager.OnConnectionLost(ex.Message);
-                        break;
-                    }
-                    else if (connectionManager.GetDisconnectStatus())
-                    {
-                        connectionManager.OnDisconnected();
-                        connectionCut = false; // This disconnect is intentional
-                        break;
-                    }
-
                     continue;
                 }
-
-                if (bytesRead == 0)
+                catch (Exception ex)
                 {
+                    Logger.Log("Disconnected from CnCNet due to a socket error. Message: " + ex.Message);
                     errorTimes++;
-
-                    if (errorTimes > 30) // TODO Figure out if this hacky check is actually necessary
-                    {
-                        failedServerIPs.Add(currentConnectedServerIP);
-                        Logger.Log("Disconnected from CnCNet.");
-                        connectionManager.OnConnectionLost("Server disconnected.".L10N("UI:Main:ServerDisconnected"));
-                        break;
-                    }
-
                     continue;
                 }
 
+                if (errorTimes > MAX_RECONNECT_COUNT)
+                {
+                    string errorMessage = FormattableString.Invariant($"Disconnected from CnCNet after {errorTimes} failed retries.");
+                    Logger.Log(errorMessage);
+                    failedServerIPs.Add(currentConnectedServerIP);
+                    connectionManager.OnConnectionLost(errorMessage);
```

Comment:
In this line, the text should be a localizable literal string

Reply:
`connectionManager.OnConnectionLost(string.Format("xxxxxxx {0} xxxx", var1));`

avoid `$"str{var}"` as such a string can't be localized

Reply:
Updated

Reply:
No. Use `L10N()` on the string to be localized.

`Logger.Log("some english text")` 
`connectionManager.OnConnectionLost("some english text".L10N("the label"));`

Reply:
> This string is never shown to the user if you follow all execution paths, should we still localize it in that case?




`connectionManager.OnConnectionLost()` shows the string to the user. So it should be localized.

On the other hand, messages in `Logger.Log` should not be localized.

Reply:
> `connectionManager.OnConnectionLost()` shows the string to the user.

The string is ignored by all current subscribers to `CnCNetManager.ConnectionLost` but I added a label anyway.